File: pygyro/initialisation/__pyccel__/initialiser_funcs.f90
module initialiser_funcs

implicit none

contains

!........................................
pure  function n0(r, CN0, kN0, deltaRN0, rp) result(Out_0001)

implicit none

real(kind=8) :: Out_0001
real(kind=8), value :: r
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp

Out_0001 = CN0 * exp(-kN0 * deltaRN0 * tanh((r - rp) / deltaRN0))
return

end function n0
!........................................

!........................................
pure  function ti(r, Cti, kti, deltaRti, rp) result(Out_0002)

implicit none

real(kind=8) :: Out_0002
real(kind=8), value :: r
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
real(kind=8), value :: rp

Out_0002 = Cti * exp(-kti * deltaRti * tanh((r - rp) / deltaRti))
return

end function ti
!........................................

!........................................
pure  function perturbation(r, theta, z, m, n, rp, deltaR, R0) result( &
      Out_0003)

implicit none

real(kind=8) :: Out_0003
real(kind=8), value :: r
real(kind=8), value :: theta
real(kind=8), value :: z
integer(kind=8), value :: m
integer(kind=8), value :: n
real(kind=8), value :: rp
real(kind=8), value :: deltaR
real(kind=8), value :: R0

Out_0003 = exp(-(r - rp) ** 2_8 / deltaR) * cos(m * theta + n * z / R0)
return

end function perturbation
!........................................

!........................................
pure  function f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti &
      ) result(Out_0004)

implicit none

real(kind=8) :: Out_0004
real(kind=8), value :: r
real(kind=8), value :: vPar
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti

Out_0004 = n0(r, CN0, kN0, deltaRN0, rp) * exp(-0.5d0 * vPar * vPar / ti &
      (r, Cti, kti, deltaRti, rp)) / Real(sqrt(2.0d0 * &
      3.14159265358979d0 * ti(r, Cti, kti, deltaRti, rp)), 8)
return

end function f_eq
!........................................

!........................................
pure  function n0deriv_normalised(r, kN0, rp, deltaRN0) result(Out_0005)

implicit none

real(kind=8) :: Out_0005
real(kind=8), value :: r
real(kind=8), value :: kN0
real(kind=8), value :: rp
real(kind=8), value :: deltaRN0

Out_0005 = -kN0 * (1_8 - tanh((r - rp) / deltaRN0) ** 2_8)
return

end function n0deriv_normalised
!........................................

!........................................
pure  function te(r, Cte, kte, deltaRte, rp) result(Out_0006)

implicit none

real(kind=8) :: Out_0006
real(kind=8), value :: r
real(kind=8), value :: Cte
real(kind=8), value :: kte
real(kind=8), value :: deltaRte
real(kind=8), value :: rp

Out_0006 = Cte * exp(-kte * deltaRte * tanh((r - rp) / deltaRte))
return

end function te
!........................................

!........................................
pure  function init_f(r, theta, z, vPar, m, n, eps, CN0, kN0, deltaRN0, &
      rp, Cti, kti, deltaRti, deltaR, R0) result(Out_0007)

implicit none

real(kind=8) :: Out_0007
real(kind=8), value :: r
real(kind=8), value :: theta
real(kind=8), value :: z
real(kind=8), value :: vPar
integer(kind=8), value :: m
integer(kind=8), value :: n
real(kind=8), value :: eps
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
real(kind=8), value :: deltaR
real(kind=8), value :: R0

Out_0007 = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) * ( &
      1_8 + eps * perturbation(r, theta, z, m, n, rp, deltaR, R0))
return

end function init_f
!........................................

!........................................
pure  subroutine init_f_flux(surface, r, theta, zVec, vPar, m, n, eps, &
      CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti, deltaR, R0)

implicit none

real(kind=8), intent(inout) :: surface(0:,0:)
real(kind=8), value :: r
real(kind=8), intent(in) :: theta(0:)
real(kind=8), intent(in) :: zVec(0:)
real(kind=8), value :: vPar
integer(kind=8), value :: m
integer(kind=8), value :: n
real(kind=8), value :: eps
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
real(kind=8), value :: deltaR
real(kind=8), value :: R0
integer(kind=8) :: i
real(kind=8) :: q
integer(kind=8) :: j
real(kind=8) :: z

do i = 0_8, size(theta,1)-1_8, 1_8
q = theta(i)
do j = 0_8, size(zVec,1)-1_8, 1_8
z = zVec(j)
surface(j, i) = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti &
      ) * (1_8 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))
end do
end do

end subroutine init_f_flux
!........................................

!........................................
pure  subroutine init_f_pol(surface, rVec, theta, z, vPar, m, n, eps, &
      CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti, deltaR, R0)

implicit none

real(kind=8), intent(inout) :: surface(0:,0:)
real(kind=8), intent(in) :: rVec(0:)
real(kind=8), intent(in) :: theta(0:)
real(kind=8), value :: z
real(kind=8), value :: vPar
integer(kind=8), value :: m
integer(kind=8), value :: n
real(kind=8), value :: eps
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
real(kind=8), value :: deltaR
real(kind=8), value :: R0
integer(kind=8) :: i
real(kind=8) :: q
integer(kind=8) :: j
real(kind=8) :: r

do i = 0_8, size(theta,1)-1_8, 1_8
q = theta(i)
do j = 0_8, size(rVec,1)-1_8, 1_8
r = rVec(j)
surface(j, i) = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti &
      ) * (1_8 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))
end do
end do

end subroutine init_f_pol
!........................................

!........................................
pure  subroutine init_f_vpar(surface, r, theta, z, vPar, m, n, eps, CN0, &
      kN0, deltaRN0, rp, Cti, kti, deltaRti, deltaR, R0)

implicit none

real(kind=8), intent(inout) :: surface(0:,0:)
real(kind=8), value :: r
real(kind=8), intent(in) :: theta(0:)
real(kind=8), value :: z
real(kind=8), intent(in) :: vPar(0:)
integer(kind=8), value :: m
integer(kind=8), value :: n
real(kind=8), value :: eps
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
real(kind=8), value :: deltaR
real(kind=8), value :: R0
integer(kind=8) :: i
real(kind=8) :: q
integer(kind=8) :: j
real(kind=8) :: v

do i = 0_8, size(theta,1)-1_8, 1_8
q = theta(i)
do j = 0_8, size(vPar,1)-1_8, 1_8
v = vPar(j)
surface(j, i) = f_eq(r, v, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) * &
      (1_8 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))
end do
end do

end subroutine init_f_vpar
!........................................

!........................................
pure  subroutine feq_vector(surface, r_vec, vPar, CN0, kN0, deltaRN0, rp &
      , Cti, kti, deltaRti)

implicit none

real(kind=8), intent(inout) :: surface(0:,0:)
real(kind=8), intent(in) :: r_vec(0:)
real(kind=8), intent(in) :: vPar(0:)
real(kind=8), value :: CN0
real(kind=8), value :: kN0
real(kind=8), value :: deltaRN0
real(kind=8), value :: rp
real(kind=8), value :: Cti
real(kind=8), value :: kti
real(kind=8), value :: deltaRti
integer(kind=8) :: i
real(kind=8) :: r
integer(kind=8) :: j
real(kind=8) :: v

do i = 0_8, size(r_vec,1)-1_8, 1_8
r = r_vec(i)
do j = 0_8, size(vPar,1)-1_8, 1_8
v = vPar(j)
surface(j, i) = f_eq(r, v, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti)
end do
end do

end subroutine feq_vector
!........................................

end module initialiser_funcs

File: pygyro/initialisation/initialiser_funcs.py
from pyccel.decorators import pure


@pure
def n0(r: 'float', CN0: 'float', kN0: 'float', deltaRN0: 'float', rp: 'float') -> 'float':
    """
    TODO
    """
    from numpy import exp, tanh
    return CN0 * exp(-kN0 * deltaRN0 * tanh((r - rp) / deltaRN0))


@pure
def Ti(r: 'float', Cti: 'float', kti: 'float', deltaRti: 'float', rp: 'float') -> 'float':
    """
    TODO
    """
    from numpy import exp, tanh
    return Cti * exp(-kti * deltaRti * tanh((r - rp) / deltaRti))


@pure
def perturbation(r: 'float', theta: 'float', z: 'float', m: 'int', n: 'int',
                 rp: 'float', deltaR: 'float', R0: 'float') -> 'float':
    """
    TODO
    """
    from numpy import exp, cos
    return exp(- (r - rp)**2 / deltaR) * cos(m * theta + n * z / R0)


@pure
def f_eq(r: 'float', vPar: 'float', CN0: 'float', kN0: 'float', deltaRN0: 'float',
         rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float') -> 'float':
    """
    TODO
    """
    from numpy import exp, sqrt, pi, real
    return n0(r, CN0, kN0, deltaRN0, rp) * exp(-0.5 * vPar * vPar / Ti(r, Cti, kti, deltaRti, rp)) \
        / real(sqrt(2.0*pi * Ti(r, Cti, kti, deltaRti, rp)))


@pure
def make_f_eq_grid(CN0: 'float', kN0: 'float', deltaRN0: 'float',
                   rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float',
                   grid_v: 'float[:]', grid_r: 'float[:]', output: 'float[:,:]'):
    """
    output array is assumed to have v as the first index
    """
    for i, vPar in enumerate(grid_v):
        for j, r in enumerate(grid_r):
            output[i, j] = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti)


@pure
def make_n0_grid(CN0: 'float', kN0: 'float', deltaRN0: 'float', rp: 'float',
                 grid_r: 'float[:]', output: 'float[:]'):
    """
    TODO
    """

    for i, r in enumerate(grid_r):
        output[i] = n0(r, CN0, kN0, deltaRN0, rp)


@pure
def n0deriv_normalised(r: 'float', kN0: 'float', rp: 'float', deltaRN0: 'float') -> 'float':
    """
    TODO
    """
    from numpy import tanh
    return -kN0 * (1 - tanh((r - rp) / deltaRN0)**2)


@pure
def Te(r: 'float', Cte: 'float', kte: 'float', deltaRte: 'float', rp: 'float') -> 'float':
    """
    TODO
    """
    from numpy import exp, tanh
    return Cte * exp(-kte * deltaRte * tanh((r - rp) / deltaRte))


@pure
def init_f(r: 'float', theta: 'float', z: 'float', vPar: 'float', m: 'int', n: 'int',
           eps: 'float', CN0: 'float', kN0: 'float', deltaRN0: 'float', rp: 'float',
           Cti: 'float', kti: 'float', deltaRti: 'float', deltaR: 'float', R0: 'float') -> 'float':
    """
    TODO
    """
    return f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) \
        * (1 + eps * perturbation(r, theta, z, m, n, rp, deltaR, R0))


@pure
def init_f_flux(surface: 'float[:,:]', r: 'float', theta: 'float[:]', zVec: 'float[:]',
                vPar: 'float', m: 'int', n: 'int', eps: 'float', CN0: 'float', kN0: 'float',
                deltaRN0: 'float', rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float', deltaR: 'float', R0: 'float'):
    """
    TODO
    """
    for i, q in enumerate(theta):
        for j, z in enumerate(zVec):
            surface[i, j] = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) \
                * (1 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))


@pure
def init_f_pol(surface: 'float[:,:]', rVec: 'float[:]', theta: 'float[:]', z: 'float',
               vPar: 'float', m: 'int', n: 'int', eps: 'float', CN0: 'float', kN0: 'float',
               deltaRN0: 'float', rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float', deltaR: 'float', R0: 'float'):
    """
    TODO
    """
    for i, q in enumerate(theta):
        for j, r in enumerate(rVec):
            surface[i, j] = f_eq(r, vPar, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) \
                * (1 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))


@pure
def init_f_vpar(surface: 'float[:,:]', r: 'float', theta: 'float[:]', z: 'float', vPar: 'float[:]',
                m: 'int', n: 'int', eps: 'float', CN0: 'float', kN0: 'float', deltaRN0: 'float',
                rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float', deltaR: 'float', R0: 'float'):
    """
    TODO
    """
    for i, q in enumerate(theta):
        for j, v in enumerate(vPar):
            surface[i, j] = f_eq(r, v, CN0, kN0, deltaRN0, rp, Cti, kti, deltaRti) \
                * (1 + eps * perturbation(r, q, z, m, n, rp, deltaR, R0))


@pure
def feq_vector(surface: 'float[:,:]', r_vec: 'float[:]', vPar: 'float[:]', CN0: 'float', kN0: 'float',
               deltaRN0: 'float', rp: 'float', Cti: 'float', kti: 'float', deltaRti: 'float'):
    """
    TODO
    """
    for i, r in enumerate(r_vec):
        for j, v in enumerate(vPar):
            surface[i, j] = f_eq(r, v, CN0, kN0, deltaRN0,
                                 rp, Cti, kti, deltaRti)
